pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java

```diff
@@ -177,4 +193,87 @@ private boolean shouldDeleteInProgressLLCSegment(String segmentName, IdealState
+            .deleteSegments(tableNameWithType, computeSegmentsToDeleteFromSegmentLineage(segmentLineage));
+
+        Assert.assertTrue(response.isSuccessful());
+
+        // Fetch available segments for the table
+        Set<String> segmentsForTable = new HashSet<>(_pinotHelixResourceManager.getSegmentsFor(tableNameWithType));
+
+        // Clean up the segment lineage
+        for (String lineageEntryId : segmentLineage.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineage.getLineageEntry(lineageEntryId);
+          if (lineageEntry.getState() == LineageEntryState.COMPLETED) {
+            // The lineage entry for 'COMPLETED' state can only be safely removed when both segmentFrom & segmentTo
+            // are all removed from the table.
+            if (Collections.disjoint(segmentsForTable, lineageEntry.getSegmentsFrom()) && Collections
+                .disjoint(segmentsForTable, lineageEntry.getSegmentsTo())) {
+              segmentLineage.deleteLineageEntry(lineageEntryId);
+            }
+          } else if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+            // Zombie lineage entry is safe to remove. This will allow the task scheduler to re-schedule the
+            // source segments to be merged again.
+            segmentLineage.deleteLineageEntry(lineageEntryId);
+          }
+        }
+
+        // Write back to the lineage entry
+        return SegmentLineageAccessHelper
+            .writeSegmentLineage(_pinotHelixResourceManager.getPropertyStore(), segmentLineage, expectedVersion);
+      });
+    } catch (Exception e) {
+      String errorMsg = String.format("Failed to clean up the segment lineage. (tableName = %s)", tableNameWithType);
+      LOGGER.error(errorMsg, e);
+      throw new RuntimeException(errorMsg, e);
+    }
+    LOGGER.info("Segment lineage metadata clean-up is successfully processed for table: {}", tableNameWithType);
+  }
+
+  /**
+   * Compute the segments that can be safely deleted based on the segment lineage.
+   *
+   * 1. The original segments can be deleted once the merged segments are successfully uploaded.
+   * 2. If the segmentReplacement operation fails in the middle, there can be a case where partial segments are
+   *    uploaded to the table. We should periodically clean up those zombie segments.
+   */
+  private List<String> computeSegmentsToDeleteFromSegmentLineage(SegmentLineage segmentLineage) {
+    if (segmentLineage != null) {
+      List<String> segmentsToDelete = new ArrayList<>();
+      for (String lineageEntryId : segmentLineage.getLineageEntryIds()) {
+        LineageEntry lineageEntry = segmentLineage.getLineageEntry(lineageEntryId);
+        if (lineageEntry.getState() == LineageEntryState.COMPLETED) {
+          // If the lineage state is 'COMPLETED', it is safe to delete all segments from 'segmentsFrom'
+          segmentsToDelete.addAll(lineageEntry.getSegmentsFrom());
```

Comment:
Can we directly remove the lineage entry here? Why do we need to wait for all merged segments also being removed?

Reply:
I initially thought that we can clean up segment + lineage metadata within a single cycle if we wait for segments to be deleted. However, I think that it's fine that the segments get deleted first and the lineage entry gets cleaned up in the next round.

Reply:
Actually, if we don't send out the routing table refresh message explicitly, the routing table rebuild will only get triggered after the segment gets deleted. So, it will always be the following sequence:

1. Retention manager deletes segments, update segment lineage metadata (one of 2 operation can happen earlier).
2. Broker side routing table updates will only happen after the segment is deleted (by EV change).

As long as we have the guarantee that the routing table gets updated after the segment is deleted, we are good.

Reply:
Good point. In this case, we should always delete the lineage entry only if the segments are already deleted (for both `COMPLETED` and `IN_PROGRESS`).
The logic should be:
If any of the segments (`segmentsFrom` for `COMPLETED` and `segmentsTo` for `IN_PROGRESS`) still exists, delete the existing segments; if they are all already deleted, delete the lineage entry.

Reply:
@Jackie-Jiang For checking whether the segment is deleted, do you think checking `getSegmentsFor(tableName)` is good enough? Or checking idealstate is preferred?

Reply:
@snleee Checking ZK metadata is fine. That can prevent leaving orphan segment ZK metadata

Reply:
Sounds good.